srml/contract/src/lib.rs

```diff
@@ -498,6 +517,52 @@ decl_module! {
 			}
 		}
 
+		/// Allows a contract to restore a tombstone by giving its storage.
+		///
+		/// The contract that want to restore (i.e. origin of the call) will compute a
+		/// tombstone with its storage and the given code_hash. If the computed tombstone
+		/// match the dest ones then dest contract is restored with the rent_allowance specified,
+		/// the origin send all its fund to the the destination and gets removed.
+		fn restore_to(
+			origin,
+			dest: T::AccountId,
+			code_hash: CodeHash<T>,
+			rent_allowance: BalanceOf<T>
+		) {
+			let origin = ensure_signed(origin)?;
+
+			let origin_contract = <ContractInfoOf<T>>::get(&origin)
+				.and_then(|c| c.get_alive())
+				.ok_or("Cannot restore from inexisting or tombstone contract")?;
+
+			let dest_tombstone = <ContractInfoOf<T>>::get(&dest)
+				.and_then(|c| c.get_tombstone())
+				.ok_or("Cannot restore to inexisting or alive contract")?;
+
+			let tombstone = <TombstoneContractInfo<T>>::new(
+				runtime_io::child_storage_root(&origin_contract.trie_id),
```

Comment:
My first thought was that this code could use a comment that this operation can be heavy, but then I realize: we might have a problem since there are no levers we can pull to change the price of this exrinsic.

Reply:
if we store the number of different key-value we can get a god way to evaluate this price (probably slightly leverage by the actual bytesize as implementation need also to hash content).

Reply:
can't we make the caller pay for a price that is function of storage size ?

Reply:
the more elements in the trie the costier it gets (also true for hashing content so storage size). It would needs some bench number to get a better idea of each relative cost.

Reply:
we only have to pay for at least compute_time. so if we only have access to storage_size we can say that it cost as if this storage size was for a number of key of `number _of_key = storage_size / minimum_value_size` and for each key: `value = maximum value size`
it seems to me

Reply:
Imagine writing a contract for restoration. You would need to fill it with required storage. After you finish, you would most likely clean up the stuff spefic to the restoration contract (i.e. access control variables, maybe just the "owner" field) and then send 
the restoration extrinsic. The extrinsic should've been sent from the contract itself. The base part is covered by gas, but then if there is this extra part it should be covered by the balance of the contract. But the problem is that the contract can't ensure that it has enough funds for that. Moreover, because the application of an extrinsic is async, there is no way to find out the outcome of the extrinsic execution. This is further worsened, that we unlocked the contract (by removing the ACL stuff), so after this moment the contract can be griefed.

Reply:
maybe then ACL should be hard-coded, but that would imply not having generic restoration contract, user will have to compile its own

Reply:
ACL can be hardcoded, but that is less than ideal. We are trying to achieve less code duplication and because of this I'd expect put_code to be rather expensive. That means that duplicating the same code but with different parameters is less desirable

Reply:
I do not see any magic solution for the acl issue but to allow restore_to to ignore an address (additional parameter): just remove it before child_root calc and add it back afterward.
For the cost of a child_root calculation, I feel like it would be worth it to count the number of child in the child trie (using a upper bound based on storage_size could really be unfavorable for contract storing big content).

Reply:
Please forget what I said about the cost calculation of `child_storage_root`, I was thinking about the scenario where the root is not already build and all content is into delta (that really is not the way it work even with eth rent eip), so since the root is already build this operation should be costless (what is costfull is the calls to ext_set_storage which simply replies to standard cost and already should take into consideration a base cost per key (and some additional cost per size of content)).

Reply:
> Can we set a fixed gas price per byte of restored storage?

Note that this function doesn't deal with gas whatsoever. So we might want to speak about charging a fee in balance units. 

We spoke about my concern with @thiolliere and it turns out that the problem that I mentioned in my last post is not that big. This is because we know the state size upfront and we can compute the balance fee upfront as well.

I think we might want to reiterate on the design and reconsider to implement this function as part of the srml-contract, not as an extrinsic.

Reply:
I agree.  `srml-contract` is the better place to put this.

Reply:
contrary to expectation compute_storage_root is not incremental.
Thus either we find a way to do a compute_child_storage_root that does only take into account changes for this child.
or we delay to next block and make use of on_initialize,
or we compute_child_storage_root without changes of the block.

I would go option 1

Reply:
I also like 2.
3 with some contract specific implementation seems a bit less good (In my opinion it would require two calls to be safe).
1 seems like a big change on internals.